### PR TITLE
Async networking (phase I)

### PR DESCRIPTION
**Pull request by ekoby**

### CODE REVIEW

File: ziti/src/main/kotlin/io/netfoundry/ziti/identity/util.kt [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+    error("no suitable key entry")
+}
+
+internal fun keystoreFromConfig(id: IdentityConfig): KeyStore {
+    val ks = KeyStore.getInstance("PKCS12")
+    ks.load(null)
+
+    val certs = readCerts(id.id.cert.replace("pem:", ""))
+    val ztAPI = URI.create(id.ztAPI)
+    val name = X500Name(certs[0].subjectDN.name).getRDNs(BCStyle.CN)[0].first.value.toASN1Primitive().toString()
```
Comment (dovholuknf) [inline ziti/src/main/kotlin/io/netfoundry/ziti/identity/util.kt:46]: position 0 is guaranteed to not be null i assume. also - is position 0 guaranteed to be 'the right' one?

File: ziti/src/main/kotlin/io/netfoundry/ziti/identity/util.kt [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+
+}
+
+internal fun loadKeystore(f: File, pwd: CharArray): KeyStore {
+    val ks = KeyStore.getInstance("PKCS12")
+    try {
+        ks.load(f.inputStream(), pwd)
+        return ks
+    } catch (ex: Exception) {
+        //w("failed to load $f as PKCS12 key store: $ex")
```
Comment (dovholuknf) [inline ziti/src/main/kotlin/io/netfoundry/ziti/identity/util.kt:71]: commented out on purpose?

File: ziti/src/main/kotlin/io/netfoundry/ziti/identity/util.kt [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+    }
+
+    try {
+        val id = IdentityConfig.load(f)
+        return keystoreFromConfig(id)
+    } catch (ex: Exception) {
+        println(ex)
```
Comment (dovholuknf) [inline ziti/src/main/kotlin/io/netfoundry/ziti/identity/util.kt:78]: println or should this be an `e` or `w`?

File: ziti/src/main/kotlin/io/netfoundry/ziti/net/nio/AsyncSocketImpl.kt [hunk 4]
```diff
@@ -0,0 +1,188 @@
...
+
+    override fun connect(host: String, port: Int) =
+        connect(InetSocketAddress(host, port), 0)
+
+    override fun connect(address: InetAddress, port: Int) = connect(
+        InetSocketAddress(address, port), 0)
+
+    override fun connect(address: SocketAddress, timeout: Int) {
+        val cf = channel.connect(address)
+        if (timeout > 0) {
```
Comment (dovholuknf) [inline ziti/src/main/kotlin/io/netfoundry/ziti/net/nio/AsyncSocketImpl.kt:78]: this should have a floor associated with it right?

File: ziti/src/main/kotlin/io/netfoundry/ziti/net/nio/AsyncSocketImpl.kt [hunk 5]
```diff
@@ -0,0 +1,188 @@
...
+                            override fun failed(exc: Throwable, f: CompletableFuture<Int>) {
+                                if (!f.isDone) {
+                                    f.completeExceptionally(exc)
+                                }
+                                inputLock.release()
+                            }
+                        })
+
+                    try {
+                        val read = rf.get(to.toLong(), TimeUnit.MILLISECONDS)
```
Comment (dovholuknf) [inline ziti/src/main/kotlin/io/netfoundry/ziti/net/nio/AsyncSocketImpl.kt:129]: same floor comment wrt timeouts. i expect it's accounted for somewhere but just pointing it out

File: ziti/src/main/kotlin/io/netfoundry/ziti/net/nio/AsyncTLSChannel.kt [hunk 6]
```diff
@@ -0,0 +1,507 @@
...
+        closed
+    }
+
+    internal val handshake = CompletableFuture<SSLSession>()
+    private var state: State
+    internal val transport: AsynchronousSocketChannel
+    internal lateinit var engine: SSLEngine
+
+    private val sslbuf = ByteBuffer.allocate(32 * 1024)
+    private val plnbuf = ByteBuffer.allocate(32 * 1024).apply { flip() }
```
Comment (dovholuknf) [inline ziti/src/main/kotlin/io/netfoundry/ziti/net/nio/AsyncTLSChannel.kt:91]: i've seen 32k in here at least 3 times now - is that worthy of a constant?

File: ziti/src/main/kotlin/io/netfoundry/ziti/identity/util.kt (hunk 1 above)
Comment (ekoby) [reply ziti/src/main/kotlin/io/netfoundry/ziti/identity/util.kt:46]: there is only one, just API is overly generic

File: ziti/src/main/kotlin/io/netfoundry/ziti/identity/util.kt (hunk 2 above)
Comment (ekoby) [reply ziti/src/main/kotlin/io/netfoundry/ziti/identity/util.kt:71]: this code was just relocated, also I don't think we need the warning there

File: ziti/src/main/kotlin/io/netfoundry/ziti/net/nio/AsyncSocketImpl.kt (hunk 4 above)
Comment (ekoby) [reply ziti/src/main/kotlin/io/netfoundry/ziti/net/nio/AsyncSocketImpl.kt:78]: timeout is checked in the base Socket before impl is called

File: ziti/src/main/kotlin/io/netfoundry/ziti/net/nio/AsyncSocketImpl.kt (hunk 5 above)
Comment (ekoby) [reply ziti/src/main/kotlin/io/netfoundry/ziti/net/nio/AsyncSocketImpl.kt:129]: same as above Socket.setSoTimeout checks the value

File: ziti/src/main/kotlin/io/netfoundry/ziti/net/nio/AsyncTLSChannel.kt [hunk 7]
```diff
@@ -0,0 +1,507 @@
...
+                    it.complete(null)
+                }
+            }
+        }
+
+        if (state == State.connecting) {
+            state = State.handshaking
+            engine.useClientMode = true
+            engine.beginHandshake()
+            continueHandshake(Handhaker(handshake, ByteBuffer.allocate(32 * 1024)))
```
Comment (dovholuknf) [inline ziti/src/main/kotlin/io/netfoundry/ziti/net/nio/AsyncTLSChannel.kt:390]: look at all them 32 * 1024!!! :)

File: ziti/src/main/kotlin/io/netfoundry/ziti/net/nio/AsyncTLSChannel.kt [hunk 8]
```diff
@@ -0,0 +1,507 @@
...
+                        continueHandshake(hs)
+                    }
+                }
+
+                SSLEngineResult.HandshakeStatus.NOT_HANDSHAKING -> {
+                    state = State.connected
+                    hs.result.complete(engine.session)
+                }
+
+                else -> TODO()
```
Comment (dovholuknf) [inline ziti/src/main/kotlin/io/netfoundry/ziti/net/nio/AsyncTLSChannel.kt:480]: anything else to add to TODO()? I don't know what the functino does but this just caught my eye

File: ziti/src/test/kotlin/io/netfoundry/ziti/net/TransportTest.kt [hunk 9]
```diff
@@ -0,0 +1,54 @@
...
+"""
+            t.write(StandardCharsets.UTF_8.encode(req))
+            val respBuf = ByteBuffer.allocate(1024)
+            t.read(respBuf, false)
+
+            respBuf.flip()
+            val resp = StandardCharsets.UTF_8.decode(respBuf)
+            println(resp)
+            val lines = resp.toString().reader().readLines()
+            assertThat(lines[0], CoreMatchers.startsWith("HTTP/1.1 200 OK"))
```
Comment (dovholuknf) [inline ziti/src/test/kotlin/io/netfoundry/ziti/net/TransportTest.kt:49]: while i suppose even getting an http 200 means 'things worked' i think it'd be better to verify some actual text?

File: ziti/src/test/kotlin/io/netfoundry/ziti/net/TransportTest.kt [hunk 10]
```diff
@@ -0,0 +1,54 @@
...
+import org.junit.Test
+import java.nio.ByteBuffer
+import java.nio.charset.StandardCharsets
+import javax.net.ssl.SSLContext
+
+class TransportTest {
+
+    @Test
+    fun testHTTP() {
+        val t = Transport.dial("tls://httpbin.org:443", SSLContext.getDefault())
```
Comment (dovholuknf) [inline ziti/src/test/kotlin/io/netfoundry/ziti/net/TransportTest.kt:31]: httpbin.org is a legit url - is this testing ziti? how do we know ziti 'worked'?

File: ziti/src/test/kotlin/io/netfoundry/ziti/net/nio/AsyncTLSChannelTest.kt [hunk 11]
```diff
@@ -0,0 +1,140 @@
...
+
+    @After
+    fun tearDown() {
+        ch.close()
+    }
+
+    @Test
+    fun connect() {
+        ch = AsyncTLSChannel.open()
+        ch.connect(InetSocketAddress("httpbin.org", 443)).get(1, TimeUnit.SECONDS)
```
Comment (dovholuknf) [inline ziti/src/test/kotlin/io/netfoundry/ziti/net/nio/AsyncTLSChannelTest.kt:52]: is the 1 the timeout? for sure some day 1 second won't be enough. for my money, i'd encourage more ... 5 or even 10 seconds. if that timeout hits - sure the build will be way slow but i think that's better than a false positive some day. also - 1 second would be a nice TIMEOUT constant seeing all the 1, TimeUnit.SECONDS on there :)

File: ziti/src/test/kotlin/io/netfoundry/ziti/net/TransportTest.kt (hunk 10 above)
Comment (ekoby) [reply ziti/src/test/kotlin/io/netfoundry/ziti/net/TransportTest.kt:31]: this is not a over-Ziti-test. transport is used to communicate to edge routers, this just tests the general networking

File: ziti/src/test/kotlin/io/netfoundry/ziti/net/TransportTest.kt (hunk 9 above)
Comment (ekoby) [reply ziti/src/test/kotlin/io/netfoundry/ziti/net/TransportTest.kt:49]: this is general networking test (connection, handshake, write request), 200 is good enough to know our request made it out there

**Pull request merged**